enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+
+As a technician with little or no OpenShift knowledge I want to be able to
+upgrade a single-node or multi-node cluster using the upgrade bundle and its
+documentation. I can't bring up any additional infrastructure at the cluster
+site, in particular I can't bring up an image registry server, neither outside
+of the cluster nor inside. I want to plug the USB stick provided by the
+engineer in one of the nodes of the cluster and have the rest of the process
+performed automatically.
+
+### Goals
+
+Provide an automated and documented mechanism that engineers managing a cluster
+can use to pin and pre-load images in order to upgrade a cluster without
+requiring a registry server.
+
+### Non-Goals
+
+It is not a goal to not require a registry server for other operations. For
+example, installing a new workload will still require a registry server.
+
+## Proposal
+
+### Workflow Description
+
+For all kinds of clusters, connected or disconnected, with or without an
+available registry server:
+
+1. The administrator of a cluster uses the OpenShift API to request an upgrade.
+
+1. The upgrade infrastructure of the cluster ensures that all the images
+required for the upgrade are pinned and pre-loaded in all the nodes of the
+cluster.
+
+In addition, for the cases were the cluster is completely disconnected or it
+isn't possible to use a registry server:
+
+1. An engineer uses the `oc adm upgrade create bundle ...` tool described in
+this enhancement to prepare the upgrade bundle containing all the artifacts
+(container images and metadata) that are required to perform the upgrade plus
+the images required for the custom workloads, and writes it to a USB stick (or
+any other suitable media) that will be handed over to the technicians
+responsible for performing the upgrades, together with documentation explaining
+how to use it.
+
+1. The technicians receive copies of the USB stick and the corresponding
+documentation.
+
+1. The technician goes to the cluster site and uses the upgrade bundle inside
+the to perform the upgrade. The documentation will ask the technician to
+plug the USB stick in one of the nodes of the cluster and provide simple
+instructions to verify that the upgraded has been applied correctly. This step
+is potentially repeated multiple times by the same technician for multiple
+clusters using the same USB stick or copies of it.
+
+Note that the upgrade bundle should not be specific for a particular cluster,
+only for the OpenShift architecture and version. Technicians should be able to
+use that package for any cluster with that architecture and version.
+
+### API Extensions
+
+There are no new object kinds introduced by this enhancement, but new fields
+will be added to existing `ClusterVersion` and `ContainerRuntimeConfig` objects.
+Those fields are described in detail in the implementation details section below,
+this is a summary of the new `CluserVersion` fields:
+
+```yaml
```

Comment:
Is every field listed here new? `DesiredUpdate` is not right?
Would you be able to convert this to go structures? It's easier to review and there are a lot of nuances that can't be picked up in YAML.

When converting to Go structures, please review https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md

If you wish to, we can review it as a PR to the openshift/api repository as well, that may be easier for you.

CC @ingvagabund in case you're interested on taking on this API review

Reply:
@JoelSpeed the `DesiredUpdate` field already exists, we are adding new fields.

I went ahead and submitted draft pull requests for the suggested API changes in the machine config operator and the cluster version operator:

- https://github.com/openshift/machine-config-operator/pull/3839
- https://github.com/openshift/api/pull/1548

Those are basically copies of what I wrote in this pull request, but a bit more oriented towards users, as what I wrote here is more oriented towards developers. There is for sure room for improvement there. Let me know if you still think it is preferable to use Go syntax here as well.

Reply:
I'll review the API via the openshift/api PR, as mentioned, there's a lot of nuance in APIs that can't be represented just in the YAML representation